src/util/mod.rs

```diff
@@ -0,0 +1,110 @@
+// person obtaining a copy of this software, associated documentation and/or
+// data (collectively the "Software"), free of charge and under any and all
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use libc::{c_int, uintptr_t};
+use ::errors::TraceMeError;
+
+// FFI prototypes.
+extern "C" {
+    fn traceme_exec_base(addr: *const uintptr_t) -> c_int;
+}
+
+/// Get the relocated virtual start address of the executable code for the current program.
+pub fn exec_base() -> Result<usize, TraceMeError> {
+    let mut addr: uintptr_t = 666;
+    match unsafe {
+        traceme_exec_base(&mut addr as *const uintptr_t)
+    } {
+        0 => Err(TraceMeError::ElfError(String::from("failed to get executable base address"))),
+        1 => Ok(addr as usize),
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::exec_base;
+
+    #[cfg(target_os = "linux")]
+    use libc::getpid;
+    #[cfg(target_os = "linux")]
+    use std::path::PathBuf;
+    #[cfg(target_os = "linux")]
+    use std::fs::File;
+    #[cfg(target_os = "linux")]
+    use std::io::BufReader;
+    #[cfg(target_os = "linux")]
+    use std::io::BufRead;
+    #[cfg(target_os = "linux")]
+    use std::env::current_exe;
+
+    /// Check we can get the exec_base() address.
+    ///
+    /// On Linux, also check this agrees with the procfs map file.
+    #[test]
+    fn test_exec_base() {
```

Comment:
This feels like it should be one test for Linux and another for not-Linux.

Reply:
Could do, but one would be a subset of the other.

Reply:
That's fine. I think it would be better than having conditional compilation inside the function (which is easily misread IMHO).

Reply:
ok